### PR TITLE
Visualisation Tool 

### PR DESCRIPTION
**Pull request by mheine94**
In this work a tool has been developed to visualize lift programs. For this purpose the user is given the opportunity to read out the data type of the input data at any point in the program and to display it visualized in a graphical user interface. Graphical representations of array, tuple, vector and scalar data types have been developed for visualization. The visualization can be adjusted via the graphical user interface by specifying values for array lengths. In addition, multidimensional arrays can be selected to be displayed as a nested array or as a matrix. The visualizations can be exported either as pixel graphics or as vector graphics.

**How to use the visualization tool**
The PrintType pattern was expanded by a TypeOutput parameter.
To get the textual output pass no parameters to PrintType.
To get a visual output one has to pass an VisualOutput object as parameter.
VisualOutput has two members:

1. render : boolean - if set to true the graphical user interface will be started and displays all datatypes that have been written to the buffer. This parameter should only be set to true if all datatypes that are desired to be displayed are already written to the buffer.

2. expression : String - if the source code of the high or low level program that contains PrintType calls is given in this parameter then it will be shown above each datatype in the visualization. The PrintType call corresponds to the given datatype is marked red in the source code above each datatype.

`def lambda = fun(
      ArrayType(Float, N), input =>
         Join() o Split(M) $ input
)`

To visualize the datatype after the split and the join function one can add PrintType calls as displayed in the following example:

`def lambda = fun(
      ArrayType(Float, N), input =>
         PrintType(VisualOutput(render = true)) o Join() o PrintType(VisualOutput()) o Split(M) $ input
)`

To display the source code one can use the parameter expression:

`var sourceCode = 
"def lambda = fun(
      ArrayType(Float, N), input =>
         PrintType(VisualOutput(render = true, expr = sourceCode)) o Join() o PrintType(VisualOutput()) o Split(M) $ input"`

`def lambda = fun(
      ArrayType(Float, N), input =>
         PrintType(VisualOutput(render = true, expr = sourceCode)) o Join() o PrintType(VisualOutput()) o Split(M) $ input
)`

In the graphical user interface one can set the values of the variables that occur in the datatypes. This is n necessary in order to display the visualization. 
After a modification was made to the variables or the "Dimensiongrouping" one can redraw the visualization by clicking the draw button or pressing enter. 
The visualization can be saved as "png" pixelgraphics or "svg" vectorgraphics by pressing the save button and by selecting the save format in the appearing save dialog.

Nested arrays can be displayed as nested arrays or as a matrix. The default mode is the nested display.

The display style can be chosen via the "Dimensiongrouping"-input field in the graphical user interface.
The display style is coded as a string.

- (1) is a single array
- (1)(2) is an array that contains arrays as elements. The array with the 2 inside the clauses represents the array on the 2nd nesting depth.
In order to display this nested array as matrix one can alter the string to (1,2)
- [ (1) , V , S ] is a tuple that has an array, vector and scalar element
- (1)[ (1)(2) , (1)(2) ] is an array that contains tuples that contain two nested array. The Nested arrays that are the tuple elements can be displayed as matrix by altering the string to:
(1)[ (1,2) , (1,2)  ]

### CODE REVIEW

File: src/main/ir/ast/PrintType.scala [hunk 1]
```diff
@@ -2,16 +2,23 @@ package ir.ast
...
 import ir.interpreter.Interpreter._
+import utils.paternoster.visualisation.TypeVisualiser
 
 /**
   * A pattern for debugging Lift code.
   * Identity function that prints the Lift type of its input.
   * Generates no OpenCL code.
   */
-case class PrintType() extends Pattern(arity = 1) {
+case class PrintType(visual: Boolean = false, render: Boolean = false,expression:String = "") extends Pattern(arity = 1) {
```
Comment (michel-steuwer) [inline src/main/ir/ast/PrintType.scala]: I would suggest creating a small algebraic data type for specifying the difference between text or visual output.
I.e. something along the lines
```
sealed trait TypeOutput
case class VisualOutput(render: Boolean, expr: String) extends TypeOutput
case object TextOutput extends TypeOutput
```

This has the advantage that it encapsulates the parameters needed for the visual output properly.
I mean it seems that the second and third argument is ignored if the first is false, but that really doesn't get clear currently (but would with the algebraic data type).

File: src/main/utils/paternoster/rendering/JavaFXRenderer.scala [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+
+  /**
+    * Method in wich the scaling of the graphics can be adjusted. It was not used in the end.
+    * The scaling was set to a fixed value to ensure readability.
+    * @param primitives All primitives.
+    * @param ctx The current context.
+    * @return The context with adjusted scaling.
+    */
+  def adjustScaling(primitives:Iterable[GraphicalPrimitive], ctx:Context):Context={
+    /*
```
Comment (michel-steuwer) [inline src/main/utils/paternoster/rendering/JavaFXRenderer.scala]: Can this code which is commented out go?

**Comment by mheine94** [comment]
I resolved the merge conflicts.
But there is an issue with this version. When I try to build the project I get the following error:

Error:(3, 13) object profiler is not a member of package lift
import lift.profiler._

The module "Profiler" is present in the project path  /lib/Profiler/ 
I don't know how i can fix this error.

**Comment by michel-steuwer** [comment]
Hi @mheine94, I think this is an issue with your IDE getting confused as the build settings were changed with the introduction of the profiler.

You should either checkout the entire repository into a fresh directory and then import it in IDEA. This should definitely work.

Alternatively, you can delete the build files of IDEA.
I believe these are stored in `target/` and `project/target/`.

**Pull request merged**